pcsx2/x86/newVif_HashBucket.h

```diff
@@ -13,87 +13,121 @@
+		u16 _pad0;
+		u32 key0;
+		u32 key1;
+		uptr value;
+	};
+
+}; // 16 bytes
+
+// 0x4000 is enough but 0x10000 allow
+// * to skip the compare value of the first double world in lookup
+// * to use a 16 bits move instead of an 'and' mask to compute the hashed key
+#define hSize 0x10000 // [usn*1:mask*1:upk*4:num*8] hash...
 
 // HashBucket is a container which uses a built-in hash function
-// to perform quick searches.
-// T is a struct data type (note: size must be in multiples of 16 bytes!)
-// hSize determines the number of buckets HashBucket will use for sorting.
+// to perform quick searches. It is designed around the nVifBlock structure
+//
 // The hash function is determined by taking the first bytes of data and
 // performing a modulus the size of hSize. So the most diverse-data should
 // be in the first bytes of the struct. (hence why nVifBlock is specifically sorted)
-template<typename T, int hSize>
 class HashBucket {
 protected:
-	SizeChain<T> mBucket[hSize];
+	std::array<nVifBlock*, hSize> m_bucket;
 
 public:
 	HashBucket() {
-		for (int i = 0; i < hSize; i++) {
-			mBucket[i].Chain	= NULL;
-			mBucket[i].Size		= 0;
-		}
+		m_bucket.fill(nullptr);
 	}
+
 	virtual ~HashBucket() throw() { clear(); }
-	int quickFind(u32 data) {
-		return mBucket[data % hSize].Size;
-	}
-	__fi T* find(T* dataPtr) {
-		u32 d = *((u32*)dataPtr);
-		const SizeChain<T>& bucket( mBucket[d % hSize] );
 
-		const __m128i* endpos = (__m128i*)&bucket.Chain[bucket.Size];
-		const __m128i data128( _mm_load_si128((__m128i*)dataPtr) );
+	__fi nVifBlock* find(const nVifBlock& dataPtr) {
+		nVifBlock* chainpos = m_bucket[dataPtr.hash_key];
 
-		for( const __m128i* chainpos = (__m128i*)bucket.Chain; chainpos<endpos; chainpos+=sizeof(T) / 16u ) {
-			// Note SSE4/AVX optimization (However it requires to only have the key in the first 16B without the pointer)
-			// tmp = xor (data128, load(chainpos))
-			// ptest tmp tmp (zf will be set if tmp == 0, i.e equality)
+		do {
+			if (chainpos->key0 == dataPtr.key0 && chainpos->key1 == dataPtr.key1)
+				return chainpos;
 
-			// This inline SSE code is generally faster than using emitter code, since it inlines nicely. --air
-			int result = _mm_movemask_ps( (cast_m128) _mm_cmpeq_epi32( data128, _mm_load_si128(chainpos) ) );
-			if( (result&0x7) == 0x7 ) return (T*)chainpos;
-		}
-		return NULL;
+			chainpos++;
+
+		} while(chainpos->startPtr != 0);
+
+		return nullptr;
 	}
-	__fi void add(const T& dataPtr) {
-		u32 d = (u32&)dataPtr;
-		SizeChain<T>& bucket( mBucket[d % hSize] );
 
-		if( (bucket.Chain = (T*)pcsx2_aligned_realloc( bucket.Chain, sizeof(T)*(bucket.Size+1), 16, sizeof(T)*bucket.Size)) == NULL ) {
+	void add(const nVifBlock& dataPtr) {
+		u32 b = dataPtr.hash_key;
+
+		u32 size = bucket_size( dataPtr );
+
+		// Warning there is an extra +1 due to the empty cell
+		if( (m_bucket[b] = (nVifBlock*)pcsx2_aligned_realloc( m_bucket[b], sizeof(nVifBlock)*(size+2), 16, sizeof(nVifBlock)*(size+1) )) == NULL ) {
```

Comment:
Since SSE isn't used anymore, maybe aligned realloc/malloc/free don't need to be used? Though it won't hurt to leave it in.